### PR TITLE
Scripts for Benchmarking CARLA, DiCE, and alibi

### PR DESCRIPTION
**Pull request by Grahams-Uncle**
Scripts for benchmarking:
- Carla Wachter
- Alibi CFProto
- DiCE DiverseCF

### CODE REVIEW

File: scripts/benchmark_proto.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    nn = nn_ohe()
+    nn.summary()
+    nn.fit(X_train, to_categorical(y_train), batch_size=256, epochs=30, verbose=0)
+
+    x_test = X_test[0].reshape((1,) + X[0].shape)
+    shape = x_test.shape
+    beta = .01
+    c_init = 1.
+    c_steps = 5
+    max_iterations = 500
```
Comment (BirkhoffG) [inline scripts/benchmark_proto.py:85]: This should be `max_iterations=1000`

**Review by BirkhoffG — CHANGES_REQUESTED** [review]
Run 1000 iterations for protocf, and test the results.

File: scripts/benchmark_proto.py (hunk 1 above)
Comment (Grahams-Uncle) [reply scripts/benchmark_proto.py]: CFProto (1000 iter):
cpu: 1007.2252881526947
gpu: 1107.2103197574615

File: scripts/benchmark_dice.py
Comment (BirkhoffG) [inline scripts/benchmark_dice.py:1]: I think this implementation is not what we implemented. We implement the "gradient" version.

You should use the example here:
http://interpret.ml/DiCE/notebooks/DiCE_getting_started.html#Explaining-a-Pytorch-model

File: scripts/benchmark_carla.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    df.to_csv("assets/adult/data.csv", index=None)
+
+    dataset = CsvCatalog(file_path="assets/adult/data.csv",
+                        continuous=continuous,
+                        categorical=categorical,
+                        immutables=immutable,
+                        target='income')
+
+    # load artificial neural network from catalog
+    model = MLModelCatalog(dataset, "linear", "pytorch", load_online=False)
```
Comment (BirkhoffG) [inline scripts/benchmark_carla.py:31]: Use `"ann"` instead of `"linear"`

**Review by BirkhoffG — CHANGES_REQUESTED** [review]

File: scripts/benchmark_dice.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+# Using method=random for generating CFs
+exp = dice_ml.Dice(d, m, method="gradient")
+
+df = df.drop('income', axis=1)
+# The number of instances for cf generation
+num_instances = 100
+
+start_time = time.time()
+print("Start...")
+dice_exp = exp.generate_counterfactuals(df.head(num_instances), total_CFs=num_instances, desired_class="opposite")
```
Comment (BirkhoffG) [inline scripts/benchmark_dice.py:63]: `total_CFs` should be 5

File: scripts/benchmark_dice.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+d = dice_ml.Data(dataframe=df,
+                 continuous_features=continuous,
+                 outcome_name='income')
+
+# Fitting a dense neural network model
+ann_model = keras.Sequential()
+ann_model.add(keras.layers.Dense(20, input_shape=(X_train.shape[1],), kernel_regularizer=keras.regularizers.l1(0.001), activation=tf.nn.relu))
+ann_model.add(keras.layers.Dense(1, activation=tf.nn.sigmoid))
+ann_model.compile(loss='binary_crossentropy', optimizer=tf.keras.optimizers.Adam(learning_rate=0.01), metrics=['accuracy'])
+ann_model.fit(X_train, y_train, validation_split=0.20, epochs=50, verbose=0, class_weight={0:1,1:2})
```
Comment (BirkhoffG) [inline scripts/benchmark_dice.py:50]: I think epochs=10 is enough.

File: scripts/benchmark_dice.py (hunk 3 above)
Comment (BirkhoffG) [inline scripts/benchmark_dice.py:63]: You should pass `X_test[:num_instances]`. You are passing the training set now.

File: scripts/benchmark_dice.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+
+df = pd.read_csv('assets/adult/data.csv')
+target = df["income"]
+
+continuous = ["age","hours_per_week"]
+categorical = ["workclass","education","marital_status","occupation","race","gender"]
+
+
+X_train, X_test, y_train, y_test = train_test_split(
```
Comment (BirkhoffG) [inline scripts/benchmark_dice.py:22]: Write this:

```python
X_train, X_test, y_train, y_test = train_test_split(
                                                    df,
                                                    target,
                                                    test_size=0.2,
                                                    shuffle=False,
                                                    stratify=target)
```

**Review by BirkhoffG — CHANGES_REQUESTED** [review]

File: scripts/benchmark_dice.py (hunk 5 above)
Comment (Grahams-Uncle) [reply scripts/benchmark_dice.py:22]: ValueError: Stratified train/test split is not implemented for shuffle=False

**Review by BirkhoffG — APPROVED** [review]

**Pull request merged**